### PR TITLE
Add server-level write batching.

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This commit changes the Server to batch writes together -- even across multiple requests. The server then periodically sends writes together to the broker by shard in a single message.

This approach uses a lot of goroutines but it allows for all points to manage the returned error and index. This simpler approach was used to get basic batching working. I'll submit a separate PR for optimizing batching to minimize goroutines.

## Notes

Batching is only implemented for raw writes. Non-raw writes only occur once for brand new series so I don't think it's worth implementing for that right now.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -42,6 +42,9 @@ const (
 
 	// DefaultShardRetention is the length of time before a shard is dropped.
 	DefaultShardRetention = 7 * (24 * time.Hour)
+
+	// DefaultFlushInterval is the time between flushing raw point data.
+	DefaultFlushInterval = 100 * time.Millisecond
```
Comment (otoolep) [inline server.go]: Definitely need to get this stuff into the config. Can be a separate PR.

File: server.go [hunk 2]
```diff
@@ -2543,6 +2577,42 @@ func (s *Server) processor(client MessagingClient, done chan struct{}) {
 	}
 }
 
+// flusher sends raw point data in batches to the broker by topic.
+func (s *Server) flusher(client MessagingClient, done chan struct{}) {
+	ticker := time.NewTicker(s.FlushInterval)
```
Comment (otoolep) [inline server.go]: Is `Ticker` better than using `After` in the `select` block? Does it matter?

Comment (benbjohnson) [inline server.go]: A `Ticker` will always fire every `x` duration whereas an `After` will fire every `x+elapsed` duration (where `elapsed` is the time taken between calls to `After`).

Comment (otoolep) [inline server.go]: Ah yes -- a subtle, but important point. I used `After` in retention policy enforcement, but the RP check-intervals are very large there relative to milliseconds, so I think the RP code is fine.

**Comment by otoolep** [comment]
Perhaps an alternative using channels would work, and we could remove the use of server locks. `WriteSeries` would create a buffer, shove all the points into the buffer (well, pointers to the points perhaps), and also tack on a "done" channel to the buffer object. `WriteSeries` sends the buffer down another channel, specially for the `Flusher`. `WriteSeries` then blocks, waiting on the "done" channel for a single message to be received.

The `Flusher` then performs the batching it wants, and sends a new message back over the "done" channel when that batch has been processed, the message containing "OK" or "ERROR" for each point in the batch. I know that a batch could contain a mix of non-raw and raw series data, but handing that is all within `Flusher`, `WriteSeries` wouldn't care. Additionally, `Flusher` could be more complex, and process *multiple* buffers at a time received over its channel, combining the buffers into an even bigger batch, signalling "done" on each buffer when the larger batch has completed.

If we want completely decoupled flow, then `WriteSeries` doesn't even both to block on its "done" channel.

Obviously I may be missing some requirement, but I like this because it means a) no goroutines in `WriteSeries` and b) no server-level locking (that I can think of).

**Comment by benbjohnson** [comment]
Closing because https://github.com/influxdb/influxdb/pull/1644 implements batching in a different way.
